src/lib.rs

```diff
@@ -1,90 +1,221 @@
 }
 
+unsafe impl Send for IcebergArrowStream {}
+
 #[repr(C)]
 pub struct ArrowBatch {
     pub data: *const u8,
     pub length: usize,
-    pub rust_ptr: *mut std::ffi::c_void,
+    pub rust_ptr: *mut c_void,
+}
+
+// Response types for async operations
+#[repr(C)]
+pub struct IcebergTableResponse {
+    result: CResult,
+    table: *mut IcebergTable,
+    error_message: *mut c_char,
+    context: *const Context,
+}
+
+unsafe impl Send for IcebergTableResponse {}
+
+impl RawResponse for IcebergTableResponse {
+    type Payload = IcebergTable;
+    fn result_mut(&mut self) -> &mut CResult {
+        &mut self.result
+    }
+    fn context_mut(&mut self) -> &mut *const Context {
+        &mut self.context
+    }
+    fn error_message_mut(&mut self) -> &mut *mut c_char {
+        &mut self.error_message
+    }
+    fn set_payload(&mut self, payload: Option<Self::Payload>) {
+        match payload {
+            Some(table) => {
+                let table_ptr = Box::into_raw(Box::new(table));
+                self.table = table_ptr;
+            }
+            None => self.table = ptr::null_mut(),
+        }
+    }
+}
+
+#[repr(C)]
+pub struct IcebergArrowStreamResponse {
+    result: CResult,
+    stream: *mut IcebergArrowStream,
+    error_message: *mut c_char,
+    context: *const Context,
+}
+
+unsafe impl Send for IcebergArrowStreamResponse {}
+
+impl RawResponse for IcebergArrowStreamResponse {
+    type Payload = IcebergArrowStream;
+
+    fn result_mut(&mut self) -> &mut CResult {
+        &mut self.result
+    }
+
+    fn context_mut(&mut self) -> &mut *const Context {
+        &mut self.context
+    }
+
+    fn error_message_mut(&mut self) -> &mut *mut c_char {
+        &mut self.error_message
+    }
+
+    fn set_payload(&mut self, payload: Option<Self::Payload>) {
+        match payload {
+            Some(stream) => {
+                self.stream = Box::into_raw(Box::new(stream));
+            }
+            None => self.stream = ptr::null_mut(),
+        }
+    }
 }
 
 #[repr(C)]
-pub enum IcebergResult {
-    IcebergOk = 0,
-    IcebergError = -1,
-    IcebergNullPointer = -2,
-    IcebergIoError = -3,
-    IcebergInvalidTable = -4,
-    IcebergEndOfStream = -5,
+pub struct IcebergBatchResponse {
+    result: CResult,
+    batch: *mut ArrowBatch,
+    error_message: *mut c_char,
+    context: *const Context,
+}
+
+unsafe impl Send for IcebergBatchResponse {}
+
+impl RawResponse for IcebergBatchResponse {
+    type Payload = Option<RecordBatch>;
+    fn result_mut(&mut self) -> &mut CResult {
+        &mut self.result
+    }
+    fn context_mut(&mut self) -> &mut *const Context {
+        &mut self.context
+    }
+    fn error_message_mut(&mut self) -> &mut *mut c_char {
+        &mut self.error_message
+    }
+    fn set_payload(&mut self, payload: Option<Self::Payload>) {
+        match payload.flatten() {
+            Some(batch) => {
+                let arrow_batch = serialize_record_batch(batch);
```

Comment:
We found a bottleneck related to serialization of the response in a single thread. More performant approach would be to do something like [here](https://github.com/RelationalAI/iceberg_rust_ffi/blob/bcd578458c246fd0f25ee79e2b39bad827cdaeac/src/lib.rs#L419-L432). However, that means that the stream created there would contain RecordBatches, which are unsafe objects. I'm fine with that, but it breaks the pattern that the serialization happens in the `set_payload`.

Would you want to include the change from the above-linked branch in this PR? Or do you want to leave a TODO comment?